lib/internal/vm/source_text_module.js

```diff
@@ -111,114 +98,167 @@ class SourceTextModule {
         'options.importModuleDynamically', 'function', importModuleDynamically);
     }
 
-    const wrap = new ModuleWrap(src, url, context, lineOffset, columnOffset);
-    wrapMap.set(this, wrap);
-    linkingStatusMap.set(this, 'unlinked');
-    wrapToModuleMap.set(wrap, this);
+    let { specifier } = options;
+    if (specifier !== undefined) {
+      validateString(specifier, 'options.specifier');
+    } else if (context === undefined) {
+      specifier = `${defaultModuleName}(${globalModuleId++})`;
+    } else if (perContextModuleId.has(context)) {
+      const curId = perContextModuleId.get(context);
+      specifier = `${defaultModuleName}(${curId})`;
+      perContextModuleId.set(context, curId + 1);
+    } else {
+      specifier = `${defaultModuleName}(0)`;
+      perContextModuleId.set(context, 1);
+    }
 
-    binding.callbackMap.set(wrap, {
+    this.#wrap = new ModuleWrap(
+      source, specifier, context,
+      lineOffset, columnOffset,
+    );
+    wrapToModuleMap.set(this.#wrap, this);
+    this.#specifier = specifier;
+    this.#context = context;
+
+    binding.callbackMap.set(this.#wrap, {
       initializeImportMeta,
-      importModuleDynamically: importModuleDynamically ? async (...args) => {
-        const m = await importModuleDynamically(...args);
-        if (isModuleNamespaceObject(m)) {
-          return m;
-        }
-        if (!m || !wrapMap.has(m))
-          throw new ERR_VM_MODULE_NOT_MODULE();
-        const childLinkingStatus = linkingStatusMap.get(m);
-        if (childLinkingStatus === 'errored')
-          throw m.error;
-        return m.namespace;
-      } : undefined,
+      importModuleDynamically: importModuleDynamically ?
+        importModuleDynamicallyWrap(importModuleDynamically) :
+        undefined,
     });
+  }
 
-    Object.defineProperties(this, {
-      url: { value: url, enumerable: true },
-      context: { value: context, enumerable: true },
-    });
+  get status() {
+    try {
+      this.#error;
+    } catch {
+      throw new ERR_VM_MODULE_NOT_MODULE();
+    }
```

Comment:
I suggest to abstract this into a private function for less repetition. That way it could be used as: `this.#getPrivateProperty(name)` where name would e.g., be `'#error'`.

Reply:
`obj['#error']` is not the same as `obj.#error`

Reply:
Uh, of course! Too bad that it seems like there's no nice way around this :disappointed:

Reply:
It's pretty ugly but it should be possible to abstract it like this:

```js
function getPrivateProperty(module, name) {
  try {
    switch (name) {
      case 'error':
        return module.#error;
      ...
    }
  } catch {
    throw new ERR_VM_MODULE_NOT_MODULE();
  }
}
```

Reply:
I think how the current code is the best we can get it. you can blame tc39 for not generalizing private fields better :shrug: